shared/gateway/api_middleware.go

```diff
@@ -0,0 +1,463 @@
+package gateway
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+	butil "github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/grpcutils"
+	"github.com/wealdtech/go-bytesutil"
+)
+
+// ApiProxyMiddleware is a proxy between an Eth2 API HTTP client and gRPC-gateway.
+// The purpose of the proxy is to handle HTTP requests and gRPC responses in such a way that:
+//   - Eth2 API requests can be handled by gRPC-gateway correctly
+//   - gRPC responses can be returned as spec-compliant Eth2 API responses
+type ApiProxyMiddleware struct {
+	GatewayAddress string
+	ProxyAddress   string
+	router         *mux.Router
+}
+
+type endpointData struct {
+	postRequest interface{}
+	getResponse interface{}
+	err         ErrorJson
+}
+
+// fieldProcessor applies the processing function f to a value when the tag is present on the field.
+type fieldProcessor struct {
+	tag string
+	f   func(value reflect.Value) error
+}
+
+// Run starts the proxy, registering all proxy endpoints on ApiProxyMiddleware.ProxyAddress.
+func (m *ApiProxyMiddleware) Run() error {
+	m.router = mux.NewRouter()
+
+	m.handleApiEndpoint("/eth/v1/beacon/genesis")
+	m.handleApiEndpoint("/eth/v1/beacon/states/{state_id}/root")
+	m.handleApiEndpoint("/eth/v1/beacon/states/{state_id}/fork")
+	m.handleApiEndpoint("/eth/v1/beacon/states/{state_id}/finality_checkpoints")
+	m.handleApiEndpoint("/eth/v1/beacon/headers/{block_id}")
+	m.handleApiEndpoint("/eth/v1/beacon/blocks")
+	m.handleApiEndpoint("/eth/v1/beacon/blocks/{block_id}")
+	m.handleApiEndpoint("/eth/v1/beacon/blocks/{block_id}/root")
+	m.handleApiEndpoint("/eth/v1/beacon/blocks/{block_id}/attestations")
+	m.handleApiEndpoint("/eth/v1/beacon/pool/attestations")
+	m.handleApiEndpoint("/eth/v1/beacon/pool/attester_slashings")
+	m.handleApiEndpoint("/eth/v1/beacon/pool/proposer_slashings")
+	m.handleApiEndpoint("/eth/v1/beacon/pool/voluntary_exits")
+	m.handleApiEndpoint("/eth/v1/node/identity")
+	m.handleApiEndpoint("/eth/v1/node/peers")
+	m.handleApiEndpoint("/eth/v1/node/peers/{peer_id}")
+	m.handleApiEndpoint("/eth/v1/node/peer_count")
+	m.handleApiEndpoint("/eth/v1/node/version")
+	m.handleApiEndpoint("/eth/v1/node/health")
+	m.handleApiEndpoint("/eth/v1/debug/beacon/states/{state_id}")
+	m.handleApiEndpoint("/eth/v1/debug/beacon/heads")
+	m.handleApiEndpoint("/eth/v1/config/fork_schedule")
+	m.handleApiEndpoint("/eth/v1/config/deposit_contract")
+	m.handleApiEndpoint("/eth/v1/config/spec")
+
+	return http.ListenAndServe(m.ProxyAddress, m.router)
+}
+
+func (m *ApiProxyMiddleware) handleApiEndpoint(endpoint string) {
```

Comment:
this function is really big and I would imagine hard to debug. Let's split it up

Reply:
I extracted several small functions.